### PR TITLE
feat(dataframe): Store arbitrary objects. Can assign compound data types

### PR DESCRIPTION
**Pull request by dustmop**

### CODE REVIEW

File: dataframe/at_indexer.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+// Hash cannot be used with AtIndexer
+func (ai *AtIndexer) Hash() (uint32, error) {
+	return 0, fmt.Errorf("unhashable: %s", ai.Type())
+}
+
+// String returns the index as a string
+func (ai *AtIndexer) String() string {
+	return fmt.Sprintf("AtIndexer(...)")
```
Comment (b5) [inline dataframe/at_indexer.go:37]: are we going to expand this output? If so I'd love a `TODO(dustmop)`

File: dataframe/at_indexer.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	return true
+}
+
+// Type returns the type as a string
+func (ai *AtIndexer) Type() string {
+	return fmt.Sprintf("%s.AtIndexer", Name)
+}
+
+// Get returns the value at the given key, which must be an int pair
+func (ai *AtIndexer) Get(key starlark.Value) (starlark.Value, bool, error) {
```
Comment (b5) [inline dataframe/at_indexer.go:51]: This is a good example of the "three audience" discussion we had on a call. Where those audiences are:

1. Starlark script authors (userland)
2. The computer
3. Go developers.

This function gives me questions for the third audience. This is an exported method on an exported type, which means it's part of the API go developers will see, but it accepts starlark values. To me we can steer the Go Developer audience in one of two directions:

1. We expect go developers need this method, and will need to construct `starlark.Value` arguments to call it.
2. We _don't_ expect Go Developers to call this method, or construct `starlark.Value`'s.

In the first scenario we might need to consider exporting helper functions for getting from go types to Starlark types. In the second we should unexport this method and any other method that accepts arguments from the `starlark` package.

I'm personally a fan of the latter, and would be in favor of taking a "progressive enhancement" approach. Deal with existing API's later and just focus on not introducing exported methods that accept `starlark` package arguments where possible. If you're into that approach we should un-export this method, and reserve an exported `Get` method if Go developers need it. It's signature would look something like:

```go
func (ai *AtIndexer) Get(i,j int) (interface{}, bool, error)
```

But I don't think we actually implement that method until the Go developer audience can articulate a case where it's needed. Might be able to audit our existing test suite for clues

File: dataframe/convert.go [hunk 3]
```diff
@@ -125,6 +124,32 @@ func toScalarMaybe(v starlark.Value) (interface{}, bool) {
 	return nil, false
 }
 
+// convert starlark value to a go native datum
+func toNativeValue(v starlark.Value) interface{} {
+	val, ok := toScalarMaybe(v)
+	if ok {
```
Comment (b5) [inline dataframe/convert.go:130]: combine into a one liner:

```suggestion
	if val, ok := toScalarMaybe(v); ok {
	
```

File: dataframe/convert.go [hunk 4]
```diff
@@ -125,6 +124,32 @@ func toScalarMaybe(v starlark.Value) (interface{}, bool) {
...
+	if dict, ok := v.(*starlark.Dict); ok {
+		m := make(map[string]interface{})
+		keys := dict.Keys()
+		for i := 0; i < len(keys); i++ {
+			key := keys[i]
+			val, _, _ := dict.Get(key)
+			m[toStr(key)] = toNativeValue(val)
+		}
+		return m
+	}
```
Comment (b5) [inline dataframe/convert.go:149]: type switch plz. This is a good function. could very well see it expand to include the universe of dataframe types

File: dataframe/series.go [hunk 5]
```diff
@@ -238,26 +261,79 @@ func (s *Series) len() int {
...
 }
 
+// SetAt assigns a go native type to the cell at position 'i'
+func (s *Series) SetAt(i int, any interface{}) error {
+	switch item := any.(type) {
+	case int:
+		if s.which == typeInt {
+			s.valInts[i] = item
+		} else {
+			return fmt.Errorf("TODO: implement SetAt(int) conversion")
```
Comment (b5) [inline dataframe/series.go:319]: 👍

**Review by b5 — APPROVED** [review]
looking great. Some small nits & one design question

File: dataframe/at_indexer.go (hunk 1 above)
Comment (dustmop) [reply dataframe/at_indexer.go:37]: AtIndexer objects are not printable in any reasonable way (python's pandas just shows the class name and memory address). I'll remove the 3 dots so it just looks opaque.

File: dataframe/at_indexer.go (hunk 2 above)
Comment (dustmop) [reply dataframe/at_indexer.go:51]: Can't change the signature of this function, it is used by the `Mapping` interface for when a user does `cell = df.at[3,4]`.

I agree that the general question of what methods are designed for golang and which are designed for starlark's engine needs some sort of convention, likely just a naming scheme. I'm still thinking of how best to do this, and consider it a work in progress. Unfortunately just implementing the interfaces that we need, in order to serve starlark, must take priority.

Comment (b5) [reply dataframe/at_indexer.go:51]: sounds good, and that's definitely the right priority. We should make things work for starlark users at the expense of complexity for go users whenever that tradeoff is necessary

File: dataframe/convert.go (hunk 4 above)
Comment (dustmop) [reply dataframe/convert.go:149]: Gah, I always forget about type switch. Will do 👍

**Pull request merged**